Release candidate

Alrighttt opened the pull request:
To do:
- [x] incorporate and security review https://github.com/SiaFoundation/core/pull/207 [ref](https://github.com/KomodoPlatform/sia-rust/blob/2b8676144873bb051446f7d14f2ccd02f6755feb/src/types/transaction.rs#L106)

- [x] implement methods in V2TransactionBuilder for atomic swap signing. Requires `impl SatisfyPolicy for AtomicSwap` [ref](https://github.com/KomodoPlatform/sia-rust/blob/2b8676144873bb051446f7d14f2ccd02f6755feb/src/types/spend_policy.rs#L140)

- [ ] Generic error handling in the dispatcher. Current error handling still has remnants of now refactored dispatcher. [ref](https://github.com/KomodoPlatform/sia-rust/blob/2b8676144873bb051446f7d14f2ccd02f6755feb/src/http/client.rs#L58C1-L78C2)

- [ ] Move the [example client](https://github.com/KomodoPlatform/sia-rust/blob/refactor_hash_types/src/http/client/example.rs) to a [rust doc example](https://doc.rust-lang.org/cargo/reference/cargo-targets.html#examples)

- [ ] Wrap all relevant unit tests with `cross_target_tests` [example](https://github.com/KomodoPlatform/sia-rust/blob/9b79ae3605d3deaa1a04076ae86530b8f84065c4/src/tests/transaction.rs#L11)

- [ ] Migrate [docker test framework](https://github.com/KomodoPlatform/komodo-defi-framework/blob/b2bc13a3b53c94c03b1081ed2621dc4d846465e7/mm2src/mm2_main/tests/docker_tests/sia_docker_tests.rs) from Komodo DeFi Framework into Sia Rust. There are tests throughout the codebase that rely on a hardcoded URL to a walletd index node. Remove these in favor of mocked API or walletd docker containers.

- [ ] Implement the remaining file contract related types for V2Transaction. This is not critical for final delivery as we can safely ignore this data in deserialization and will never do serialization. At a minimum, replace `todo!()`s with error handling or empty data. [ref1](https://github.com/KomodoPlatform/sia-rust/blob/2b8676144873bb051446f7d14f2ccd02f6755feb/src/types/transaction.rs#L677) [ref2](https://github.com/KomodoPlatform/sia-rust/blob/2b8676144873bb051446f7d14f2ccd02f6755feb/src/types/transaction.rs#L721)

Alrighttt committed:
remove kdf as dep

Alrighttt committed:
use rustls for reqwest on linux target

Alrighttt committed:
align dep versions to kdf

Alrighttt committed:
remove debugging code

Alrighttt committed:
fix unit test

Alrighttt committed:
move wasm client tests

Alrighttt committed:
set cdylib in Cargo.toml

Alrighttt committed:
don't use tokio for wasm test env

Alrighttt committed:
fix wasm tx signing

Alrighttt committed:
fix tokio tests

Alrighttt committed:
Cargo.lock

Alrighttt committed:
move hash to sub-module of types

Alrighttt committed:
new minimal hash module

Alrighttt committed:
publicly expose H256 in types module

Alrighttt committed:
remove PrefixedH256 from types

Alrighttt committed:
add FIXME comment regarding pub exports

Alrighttt committed:
temp fix for Currency type export

Alrighttt committed:
fix H256 usage in various unit tests

Alrighttt committed:
fix H256 usage in more unit tests

Alrighttt committed:
add macros module; remove hash from root exposed modules

Alrighttt committed:
add FIXME comment regarding lib structure

Alrighttt committed:
remove PrefixedH256 type

Alrighttt committed:
rename H256->Hash256 to match Go code

Alrighttt committed:
change cross_target_test macro to process multiple unit tests

Alrighttt committed:
add FIXME comment to Cargo.toml re: missing info fields

Alrighttt committed:
fix wasm client error unit test

Alrighttt committed:
cargo fmt

Alrighttt committed:
remove unneccesary pub export

Alrighttt committed:
remove unnecessary trait derives

Alrighttt committed:
add doc comment for cross_target_tests macro

Alrighttt committed:
add hash unit tests; cargo fmt

Alrighttt committed:
cargo fmt

Alrighttt committed:
derive Eq for Hash256

Alrighttt committed:
remove PrefixedSignature type; move Signature type to own module

Alrighttt committed:
Remove PrefixedSignature entirely; remove related serde_as remnants

Alrighttt committed:
removed derive(Copy) from Signature

Alrighttt committed:
removed derive(Copy) from Signature again

Alrighttt committed:
fix Signature import paths

Alrighttt committed:
validate signature r Point

Alrighttt committed:
signature unit tests

Alrighttt committed:
fix hash unit test

Alrighttt committed:
import curve25519-dalek; already in dep tree

Alrighttt committed:
more verbose ParseHashError

Alrighttt committed:
move PublicKey and KeyPair types to module - breaking change for KDF imports

Alrighttt committed:
rename ScoidH256 -> SiacoinOutputId

Alrighttt committed:
fix SiacoinOutputID symbol rename

Alrighttt committed:
fix typo

Alrighttt committed:
fix various imports in tests

Alrighttt committed:
move SiacoinOutputID to transaction.rs; remove serde_as hack, use type directly

Alrighttt committed:
remove unused import

Alrighttt committed:
remove comment no longer applicable; rename symbol

Alrighttt committed:
fix tests; remove unused import

Alrighttt committed:
remove unused trait impl

Alrighttt committed:
rename symbol HexArray64->Leaf

Alrighttt committed:
cargo fmt

Alrighttt committed:
remove Copy impl from Hash256

Alrighttt committed:
fix errors caused by remove of Hash256 Copy; add FIXME comment

Alrighttt committed:
move Leaf to transaction module

Alrighttt committed:
wrap tests in mod

Alrighttt committed:
remove PublicKey Deref impl

Alrighttt committed:
remove PrefixedPublicKey type

Alrighttt committed:
impl required PublicKey traits; better KeypairError handling

Alrighttt committed:
remove unused impl

Alrighttt committed:
remove Copy impl from PublicKey

Alrighttt committed:
remove unused PublicKey method

Alrighttt committed:
remove .public() method from PublicKey

Alrighttt committed:
refactor Keypair type; impl Verifier, Signer

Alrighttt committed:
cargo fmt

Alrighttt committed:
move Currency public export path

Alrighttt committed:
remove SpendPolicyHelper type and serde_with dep

Alrighttt committed:
cargo fmt

Alrighttt committed:
majorly reorganize pub exports

Alrighttt committed:
remove standalone v1_standard_address_from_pubkey; add Address method instead

Alrighttt committed:
remove now irrelevant comment

Alrighttt committed:
fix wasm test warnings

Alrighttt committed:
wrap blake2b unit tests in cross_target_tests

Alrighttt committed:
wrap encoder tests in cross_target_tests

Alrighttt committed:
wrap tests/encoding in cross_target_tests

Alrighttt committed:
wrap tests/serde in cross_target_tests; fix indentation

Alrighttt committed:
wrap tests/transaction in cross_target_tests

Alrighttt committed:
wrap tests/spend_policy in cross_target_tests

Alrighttt committed:
replace standalone atomic_swap functions with SpendPolicy methods

Alrighttt committed:
init atomic_swap mod WIP

Alrighttt committed:
init AtomicSwap types and tests

Alrighttt committed:
atomic swap component unit test

Alrighttt committed:
edit atomic swap comments

Alrighttt committed:
remove pub export of policy from AtomicSwap and AtomicSwapComponent wrappers; may revert this at some point

Alrighttt committed:
remove unnecessary generalization

Alrighttt committed:
remove PhantomData pattern

Alrighttt committed:
introduce IsValidatedSpendPolicy trait

Alrighttt committed:
fix doc comment

Alrighttt committed:
generalize IsValidatedPolicy to support SatisfiedPolicy

Alrighttt committed:
add opaque SpendPolicy path

Alrighttt committed:
fix misnomer; change "http"->"transport"

Alrighttt committed:
fix atomic_swap compilation - this file is likely to be fully refactored to reduce complexity

borngraced commented:
working on task 3 && 4

laruh commented:
could you add `.idea` to gitignore?

laruh commented on src/types/hash.rs:
```diff
@@ -0,0 +1,230 @@
+}
+
+impl fmt::LowerHex for Hash256 {
+    fn fmt(&self, f: &mut ::core::fmt::Formatter) -> ::core::fmt::Result {
```
You can avoid using `::core::` here, similar to the other implementations

laruh commented on src/transport/client/example.rs:
You can leave notes at the top of the file as file-level doc comment
```
//! THIS FILE MUST NOT INCLUDED IN WORKSPACE
//! **Some Broken Code to Ensure the Build Fails if this File is Included in the Workspace**
```

laruh commented on src/types/specifier.rs:
Suggest to improve `define_byte_array_const` macro like this:
```rust
use static_assertions::const_assert;

macro_rules! define_byte_array_const {
    ($name:ident, $size:expr, $value:expr) => {
        pub const $name: [u8; $size] = {
            // Array size must be at least as large as the value length
            const_assert!($size >= $value.len());

            let mut arr = [0u8; $size];
            let bytes = $value.as_bytes();
            let mut i = 0;
            while i < bytes.len() && i < $size {
                arr[i] = bytes[i];
                i += 1;
            }
            arr
        };
    };
}
```
const_assert! ensures that $size is greater than or equal to $value.len() at compile time. This means if there's a violation, the compile will fail, preventing unsafe operations.

Cargo lock already contains `static_assertions`, so its ok to add `static_assertions` as a dep. It will be just added in sia-rust `[[package]]`
https://github.com/KomodoPlatform/sia-rust/blob/9b79ae3605d3deaa1a04076ae86530b8f84065c4/Cargo.lock#L1309-L1313

laruh reviewed (changes requested):
Thanks for the PR! there are the first notes

Alrighttt committed:
add address and v1_address method to PublicKey

Alrighttt committed:
add unlock_condition helper method to SpendPolicy

Alrighttt committed:
reintroduce Keypair::public method; add doc comment for Keypair

Alrighttt committed:
use Keypair::public method instead of direct ref to struct member

Alrighttt committed:
add temp bandaid hack to allow KDF compilation; the impl is unused but required due to derive(Clone) on KDF's PrivKeyPolicy<T>

Alrighttt committed:
remove placeholder Clone impl

Alrighttt committed:
add Keypair::private method

Alrighttt committed:
use native rust ssl for all targets

Alrighttt committed:
use native rust ssl for all targets; linux target would require libssl-dev system dep otherwise

Alrighttt committed:
remove unnecessary new type wrapper

Alrighttt committed:
impl Clone for all endpoint request and response types

Alrighttt committed:
remove GetAddressUtxosResponse type alias

Alrighttt committed:
add Currency doc comment and TODO; add SiacoinElement doc comment

Alrighttt committed:
add api/debug/mine endpoint

Alrighttt committed:
edit endpoint doc comments

Alrighttt committed:
add clarification to debug comment

Alrighttt committed:
move ApiClientHelpers to its own module

Alrighttt committed:
add Address::from_public_key

Alrighttt committed:
make Currency::ZERO const pub

Alrighttt committed:
add ApiClientHelpers::select_unspent_outputs

Alrighttt committed:
impl FromStr for PublicKey in keypair.rs

Alrighttt committed:
doc comment typo

Alrighttt committed:
use public_key impl of Verifier instead of duplicate impl on Keypair

Alrighttt committed:
split KeypairError into PublicKeyError and PrivateKeyError

Alrighttt committed:
impl verify signature method for PublicKey

Alrighttt committed:
pub expose PublicKeyError and PrivateKeyError in types module

Alrighttt committed:
add SignatureError::VerifyFailed variant; edit display impls of SignatureError

Alrighttt committed:
add Signature::verify method

Alrighttt committed:
derive V1TransactionSansSigs Deref; remove manual Deref impl

Alrighttt committed:
derive arithmetic traits for Currency type via derive_more crate

Alrighttt committed:
add Currency::COIN const

Alrighttt committed:
remove From<u128> and From<i32> for Currency

Alrighttt committed:
add comment re: TxBuilder input signing

Alrighttt committed:
update V2TransactionBuilder methods to use ref instead for easier for loop/iter handling

Alrighttt committed:
add From<tuple> for SiacoinOutput helpers

Alrighttt committed:
add Currency::DUST const placeholder

Alrighttt committed:
make ApiClientHelpersError pub

Alrighttt committed:
simplify ApiClientHelpersError::SelectOutputs display impl

Alrighttt committed:
impl Clone, Debug for V2TransactionBuilder

Alrighttt committed:
make V2TransactionBuilder setters consume self instead of mut ref

Alrighttt committed:
switch V2TransactionBuilder setters back to using mut ref instead of consuming self

Alrighttt committed:
add ApiClientHelpers::fund_tx_single_source WIP

Alrighttt committed:
uncomment code

Alrighttt committed:
cargo fmt; Cargo.lock

Alrighttt committed:
fix V2TransactionBuilder::build so consumer can correctly chain it with other methods

mariocynicys commented:
Grabbed task 5: Wrap all relevant unit tests with cross_target_tests [example](https://github.com/KomodoPlatform/sia-rust/blob/9b79ae3605d3deaa1a04076ae86530b8f84065c4/src/tests/transaction.rs#L11)

mariocynicys commented:
Grabbed task 3: Generic error handling in the dispatcher. Current error handling still has remnants of now refactored dispatcher. [ref](https://github.com/KomodoPlatform/sia-rust/blob/2b8676144873bb051446f7d14f2ccd02f6755feb/src/http/client.rs#L58C1-L78C2)

Alrighttt committed:
fix Currency::COIN const definition

Alrighttt committed:
remove incomplete "SatisfyPolicy" trait - will explore this generalized concept in the future if resources become available

Alrighttt committed:
fix transaction unit tests compilation

Alrighttt committed:
add FeePolicy type to V2TransactionBuilder; doc comment regarding txfee rate calc; merge `impl V2TransactionBuilder` blocks

Alrighttt committed:
change Preimage to fixed sized array instead of Vec<u8>; accounts for Sia Core validation changes

Alrighttt committed:
fix transaction encoding tests to account for Vec<u8> -> [u8;32] preimage change

Alrighttt committed:
again fix transaction encoding tests to account for Vec<u8> -> [u8;32] preimage change

Alrighttt committed:
fix ApiClientHelpers::select_unspent_outputs doc comment

Alrighttt committed:
add brief doc comment for type Preimage

Alrighttt committed:
typo in doc comment

mariocynicys commented:
Grabbed task 6: Migrate [docker test framework](https://github.com/KomodoPlatform/komodo-defi-framework/blob/b2bc13a3b53c94c03b1081ed2621dc4d846465e7/mm2src/mm2_main/tests/docker_tests/sia_docker_tests.rs) from Komodo DeFi Framework into Sia Rust. There are tests throughout the codebase that rely on a hardcoded URL to a walletd index node. Remove these in favor of mocked API or walletd docker containers.

Alrighttt committed:
update fund_tx_single_source txfee handling; edit doc comment of fund_tx_single_source

Alrighttt committed:
impl TryFrom<&[u8]> for Hash256

Alrighttt committed:
change Preimage from type alias to newtype struc

Alrighttt committed:
add enum PreimageError

impl TryFrom<&[u8]> for Preimage and From<Preimage> for Vec<u8>

Alrighttt committed:
remove redundant `impl TryFrom<&str> for Hash256`. FromStr is implemented for Hash256.

Alrighttt committed:
add verbosity to ParseHashError display impls

Alrighttt committed:
Add TransactionId type alias;

Add SiacoinOutputId::new

impl Encodable for SiacoinOutputId

derive From, Into for SiacoinOutputId instead of explicit impl

Alrighttt committed:
Fix test compilation failure due to removal of `impl TryFrom<&str> for Hash256`; Use from_str instead.

add unit test test_siacoin_output_id_new

derive Default for Preimage and fix related tests

Alrighttt committed:
remove unused `impl Encodable for SiacoinOutputId`

Alrighttt committed:
Use u32 instead of u64 for output index.

Indexing via tx.siacoin_outputs[vout_index] may be a common pattern. WASM usize is u32 which would require doing lossy u64->u32

Alrighttt committed:
add ApiClientHelpers::utxo_from_txid helper

Alrighttt committed:
make SpendPolicy::atomic_swap take references for convenience

Alrighttt committed:
expose pub UtxoFromTxidError via client module

Alrighttt committed:
add dev comment TODO re: V2TransactionBuilder helpers

Alrighttt committed:
fix lossy usize comparison

Alrighttt committed:
add V2TransactionBuilder::satisfy_atomic_swap_success

Alrighttt committed:
set a Currency::DEFAULT_FEE const

change Currency::DUST to 1 hasting

Alrighttt committed:
refactor ApiClientHelpers error handling so allow single HelperError type

Alrighttt committed:
remove UtxoFromTxidError pub export from helpers module. UtxoFromTxidError is now wrapped in HelperError

Alrighttt committed:
fix tests compilation after SpendPolicy::atomic_swap method arguments were changed to take refs

Alrighttt committed:
remove incomplete atomic_swap module; move atomic swaps comment block to spend_policy module

Alrighttt committed:
remove atomic_swap module from types module

Alrighttt committed:
use Currency directly for arithmetic in ApiClientHelpers::select_unspent_outputs instead of converting to/from u128

Alrighttt committed:
fix various cargo clippy warnings/errors

Alrighttt committed:
fix spend_policy dev comments re: atomic swap SpendPolicy

Alrighttt committed:
Make SpendPolicy atomic swap helpers more verbose - add doc comment regarding arguments

Alrighttt committed:
add ApiClientHelpers::broadcast_transaction method

Alrighttt committed:
fix Preimage serialization - fixes SatisfiedPolicy serde

Alrighttt committed:
add V2TransactionBuilder::satisfy_atomic_swap_refund and V2TransactionBuilderError

Alrighttt committed:
add ApiClientHelpers::get_address_events and corresponding errors

Alrighttt committed:
add ApiClentHelpers::get_median_timestamp

add partial impl of api/consensus/tipstate walletd endpoint

Alrighttt committed:
remove commented dead code

Alrighttt committed:
add api/consensus/index/{height} and WIP api/consensus/updates/{height}::{hash} walletd endpoints

Alrighttt committed:
rename macros module to utils

add deserialize_null_as_empty_vec to utils module

Alrighttt committed:
delete old macros module

Alrighttt committed:
add consensus_updates module

Alrighttt committed:
remove consensus updates types from endpoints module - moved to dedicated module

Alrighttt committed:
Edit HelperError display impls

add HelperError::FindWhereUtxoSpent

Alrighttt committed:
finish implementing ApiClientHelpers::find_where_utxo_spent

Alrighttt committed:
add newly added "confirmations" field to Event type

add a FIXME dev comment regarding Deserialize impl of Event

Alrighttt committed:
add dev comment TODO re: EmptyResponse type

Alrighttt committed:
fix txpool endpoint response

Alrighttt committed:
fix various cargo doc hyperlinks

Alrighttt committed:
fix expected case for test_satisfied_policy_encode_hash_empty unit test

Alrighttt committed:
fix SpendPolicy::Hash encoding to account for fixed sized preimages change

Prior to this change, preimages included in SatisfiedPolicy were Vec<u8>

They are now [u8; 32]

Alrighttt committed:
remove id field from StateElement

introduce id field to SiafundElement and SiacoinElement

remove scoid: prefix for serde of SiacoinOutputId

change merkle_proof field of StateElement from Option<Vec<Hash256>> to Vec<Hash256> - serde(default) should handle both `null` and `[]` as empty Vec<>

Leave FIXME doc comment regarding the need for additional unit tests for these changes

Alrighttt committed:
change FileContractID from type alias to newtype pattern - this is not strictly needed but it better matches the Go implementation

Alrighttt committed:
fix SiacoinElement encoding

Alrighttt committed:
derive Default for FileContractID and SiacoinOutputId

Alrighttt committed:
fix various compilation errors in transaction unit tests related to StateElement consensus changes from Sia Go - removed id field from StateElement

Alrighttt committed:
fix SiacoinElement id field serde

Alrighttt committed:
add capacity field to V2FileContract

fix V2FileContractElement encoding

Alrighttt committed:
fix various transaction unit tests corresponding to rust_port_test.go inside komodoplatform/sia-core repo

Alrighttt committed:
refactor Hash256 serde to account for Sia Go SiaFoundation/core/pull/224

Prior to this change Hash256 was represented as "h:<32 byte hex string>"

It is now represented as a plain hex string with no prefix

Alrighttt committed:
remove now irrelevant dev comment

Alrighttt committed:
remove h: prefix from all unit tests

Alrighttt committed:
refactor Address serde to remove addr: prefix

Alrighttt committed:
remove addr: prefix from all Address instances in unit tests

Alrighttt committed:
fix merkle_proof field in StateElement serde - walletd is capable of providing a JSON that does not include this field, includes it as `null` or includes it as empty array

Alrighttt committed:
add more explitcit unit test for StateElement serde

Alrighttt committed:
update serde test to account for prefix removals and state element changes

Alrighttt committed:
rename symbol BlockID -> BlockId

Alrighttt committed:
derive From<BlockId> for Hash256 and vice versa instead of manually implementings

Alrighttt committed:
remove unused import

Alrighttt committed:
remove bid: prefix from BlockId serde

derive Display instead of manually implementing for BlockId

Alrighttt committed:
update SiacoinElement unit test to account for type changes

Alrighttt committed:
refactor Signature serde to remove sig: prefix

refactor SignatureError, PublicKeyError, KeypairError to match standard error handling pattern

remove PrivateKeyError

Alrighttt committed:
use Signature::try_from instead of now removed Signature::from_bytes

Alrighttt committed:
update serde tests v2ContractResolution event expected case JSON

Alrighttt committed:
manually impl Serialize Deserialize for Hash256 for better error messages

Alrighttt committed:
remove prefix from varoius unit tests

Alrighttt committed:
fix V2FileContractElement field name serde

Alrighttt committed:
ignore several tests that rely on hardcoded walletd server

Alrighttt committed:
fix expected case in various events unit tests

Alrighttt committed:
cargo clippy

Alrighttt committed:
add explicit ArbitraryData type and implement serde

Prior to this change, ArbitraryData was being serialized like [0, 1, 2, 3] which walletd accepted just fine. However, walletd will itself serialize this as a base64 string. This went unnoticed until swaps with the newly added UUID inside arbitrary data field were performed.

Alrighttt committed:
add ArbitraryData roundtrip serde unit test

Alrighttt committed:
derive From, Into for ArbitraryData newtype

Alrighttt committed:
fix ConsensusTipstateResponse serde

Alrighttt committed:
allow unknown fields in V2Transaction serde because walletd will sometimes include "id" field

Alrighttt committed:
fix trailing ? on endpoint Urls that have query params but none are provided

Alrighttt committed:
increase native client timeout from 10 seconds to 30

Alrighttt committed:
add ApiClientHelpers::get_unconfirmed_transaction and dev comments related to ApiClientHelpers::get_transaction

Alrighttt committed:
Merge pull request #9 from KomodoPlatform/refactor/remove-serde-prefixes

Refactor/remove serde prefixes
